### PR TITLE
IX: Add multi-format ad unit support

### PR DESCRIPTION
**Pull request by liliana-sortable**
This is a port of https://github.com/prebid/prebid-server/pull/2315

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/ix/IxBidder.java [hunk 1]
```diff
@@ -290,9 +292,28 @@ private static Response mergeNativeImpTrackers(Response response, List<EventTrac
                 .build();
     }
 
-    private static BidType getBidType(String impId, List<Imp> imps) {
+    private static BidType getBidType(Bid bid, List<Imp> imps) {
+        Integer mType = bid.getMtype();
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/ix/IxBidder.java:296]: According to project style, we prefer to use final keyword on variables declaration

File: src/main/java/org/prebid/server/bidder/ix/IxBidder.java [hunk 2]
```diff
@@ -290,9 +292,28 @@ private static Response mergeNativeImpTrackers(Response response, List<EventTrac
...
+        BidType bidType = mType != null ? switch (bid.getMtype()) {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/ix/IxBidder.java:297]: Seems bidType resolving logic can be moved to separate function

File: src/main/java/org/prebid/server/bidder/ix/IxBidder.java [hunk 3]
```diff
@@ -290,9 +292,28 @@ private static Response mergeNativeImpTrackers(Response response, List<EventTrac
...
+            case 2 -> BidType.video;
+            case 3 -> BidType.audio;
+            case 4 -> BidType.xNative;
+            default -> null;
+        } : null;
+        if (bidType != null) {
+            return bidType;
+        }
+
+        Optional<String> prebidType = Optional.ofNullable(bid.getExt())
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/ix/IxBidder.java:308]: final

Comment (SerhiiNahornyi) [reply src/main/java/org/prebid/server/bidder/ix/IxBidder.java:308]: If you will move logic from 315 line to separate method, then you code will be easily simplified to
```java
return Optional.ofNullable(bid.getExt())
...
.orElseGet(()->methodWithLogicFrom315Line());
```

File: src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java [hunk 4]
```diff
@@ -636,6 +636,68 @@ public void makeBidsShouldReturnValidAdmIfNativeIsPresentInImpAndAdm12() throws
                 .containsExactly(mapper.writeValueAsString(expectedNativeResponse));
     }
 
+    @Test
+    public void makeBidsShouldReturnCorrectTypeMTypeInResponse() throws JsonProcessingException {
```
Comment (SerhiiNahornyi) [inline src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:640]: makeBidsShouldReturnVideoBidIfMtypeIsTwo

Comment (SerhiiNahornyi) [reply src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:640]: Also pls create test for every mtype possible value

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/ix/IxBidder.java [hunk 5]
```diff
@@ -290,8 +292,32 @@ private static Response mergeNativeImpTrackers(Response response, List<EventTrac
...
+        if (bidType != null) {
+            return bidType;
+        }
+
+        return Optional.ofNullable(bid.getExt())
+                .map(ext -> ext.get("prebid"))
+                .map(prebid -> prebid.get("type"))
+                .map(JsonNode::asText)
+                .map(BidType::fromString)
+                .orElse(getBidTypeFromImp(imps, bid.getImpid()));
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/ix/IxBidder.java:306]: If you are using `orElse` argument value will always be calculated, to had it calculated only if really needed
pls use (.orElseGet(() -> getBidTypeFromImp(imps, bid.getImpid())))

File: src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java [hunk 6]
```diff
@@ -636,6 +636,149 @@ public void makeBidsShouldReturnValidAdmIfNativeIsPresentInImpAndAdm12() throws
...
+                mapper.writeValueAsString(
+                        givenBidResponse(
+                                bidBuilder -> bidBuilder
+                                        .impid("123")
+                                        .ext(mapper.createObjectNode()
+                                                .set("prebid", mapper.createObjectNode().put("type", "video"))))));
+
+        // when
+        final Result<List<BidderBid>> result = ixBidder.makeBids(httpCall, null);
+        // then
```
Comment (SerhiiNahornyi) [inline src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:768]: Pls make 768 line empty

File: src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java [hunk 7]
```diff
@@ -636,6 +636,149 @@ public void makeBidsShouldReturnValidAdmIfNativeIsPresentInImpAndAdm12() throws
...
+                                .video(video).build()))
+                        .build(),
+                mapper.writeValueAsString(
+                        givenBidResponse(
+                                bidBuilder -> bidBuilder
+                                        .impid("123")
+                                        .mtype(4))));
+
+        // when
+        final Result<List<BidderBid>> result = ixBidder.makeBids(httpCall, null);
```
Comment (SerhiiNahornyi) [inline src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:739]: We prefer to have empty line between`given` `when` `then` blocks

Comment (SerhiiNahornyi) [reply src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:739]: Also pls check for similar occurences

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/ix/IxBidder.java [hunk 8]
```diff
@@ -290,8 +292,32 @@ private static Response mergeNativeImpTrackers(Response response, List<EventTrac
...
+        if (bidType != null) {
+            return bidType;
+        }
+
+        return Optional.ofNullable(bid.getExt())
+                .map(ext -> ext.get("prebid"))
+                .map(prebid -> prebid.get("type"))
+                .map(JsonNode::asText)
+                .map(BidType::fromString)
+                .orElseGet(() -> getBidTypeFromImp(imps, bid.getImpid()));
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/ix/IxBidder.java:306]: What about to split different sources?
```
    private static BidType getBidType(Bid bid, List<Imp> imps) {
        return getBidTypeFromMtype(bid.getMtype())
                .or(() -> getBidTypeFromExtPrebidType(bid.getExt()))
                .orElseGet(() -> getBidTypeFromImp(imps, bid.getImpid()));
    }

    private static Optional<BidType> getBidTypeFromMtype(Integer mType) {
        final BidType bidType = mType != null ? switch (mType) {
            case 1 -> BidType.banner;
            case 2 -> BidType.video;
            case 3 -> BidType.audio;
            case 4 -> BidType.xNative;
            default -> null;
        } : null;

        return Optional.ofNullable(bidType);
    }

    private static Optional<BidType> getBidTypeFromExtPrebidType(ObjectNode bidExt) {
        return Optional.ofNullable(bidExt)
                .map(ext -> ext.get("prebid"))
                .map(prebid -> prebid.get("type"))
                .map(JsonNode::asText)
                .map(BidType::fromString);
    }

...
```

File: src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java [hunk 9]
```diff
@@ -636,6 +636,154 @@ public void makeBidsShouldReturnValidAdmIfNativeIsPresentInImpAndAdm12() throws
...
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .extracting(BidderBid::getBid)
+                .extracting(Bid::getExt)
+                .extracting(ext -> ext.get("prebid"))
+                .extracting(node -> mapper.treeToValue(node, ExtBidPrebid.class))
+                .extracting(ExtBidPrebid::getType)
+                .containsExactly(BidType.video);
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:781]: Why do we need to assert our input?

Comment (CTMBNara) [reply src/test/java/org/prebid/server/bidder/ix/IxBidderTest.java:781]: Only the adapter can affect this result. If logic appears that somehow changes this object, the corresponding test will be written.

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**